### PR TITLE
Avoid gated feature checking unconfigured expanded items

### LINKED ISSUE #32840: Macro-expanded unconfigured items are gated feature checked
Macro-expanded unconfigured items are gated feature checked, but ordinary unconfigured items are not. For example, the following should compile, ...
```rust
#![cfg_attr(attr, feature(rustc_attrs))]

macro_rules! mac {
    {} => {
        #[cfg(attr)]
        mod m {
            #[rustc_error] // ... but if not(attr), this complains about the lack of the rustc_attrs feature
            fn f() {}
        }
    }
}

mac! {}
```

If we manually expand `mac! {}`, it always compiles:
```rust
#![cfg_attr(attr, feature(rustc_attrs))]

#[cfg(attr)]
mod m {
    #[rustc_error]
    fn f() {}
}
```

### PR DESCRIPTION
**Pull request by jseyfried**
Avoid gated feature checking unconfigured macro-expanded items (fixes #32840).
Unconfigured items that are not macro-expanded are already not gated feature checked.
r? @nrc

### CODE REVIEW

File: src/librustc_driver/driver.rs [hunk 1]
```diff
@@ -662,6 +647,19 @@ pub fn phase_2_configure_and_expand(sess: &Session,
...
+        sess.track_errors(|| {
+            let features = syntax::feature_gate::check_crate(sess.codemap(),
+                                                             &sess.parse_sess.span_diagnostic,
+                                                             &krate,
+                                                             &attributes,
+                                                             sess.opts.unstable_features);
+            *sess.features.borrow_mut() = features;
+        })
+    })?;
+
```
Comment (LeoTestard) [inline src/librustc_driver/driver.rs]: There is already a second feature gate checking pass just a bit later in the same function (“complete gated feature checking 2”). I'm not sure this first pass is necessary at tall. I considered removing it as part of #32791 but then thought I'd do a separate PR.

Comment (jseyfried) [inline src/librustc_driver/driver.rs]: Good point, I'll remove it.

File: src/libsyntax/ext/expand.rs [hunk 2]
```diff
@@ -309,9 +309,7 @@ macro_rules! with_exts_frame {
 // When we enter a module, record it, for the sake of `module!`
 pub fn expand_item(it: P<ast::Item>, fld: &mut MacroExpander)
                    -> SmallVector<P<ast::Item>> {
-    let it = expand_item_multi_modifier(Annotatable::Item(it), fld);
-
-    expand_annotatable(it, fld)
+    expand_annotatable(Annotatable::Item(it), fld)
```
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: `expand_annotatable` begins with a call to `expand_item_multi_modifier`, which is idempotent (cf #21052).
This is unrelated to the rest of the PR.

**Comment by jseyfried** [comment]
cf #25544

**Comment by nikomatsakis** [comment]
Finally discussed in lang-team meeting. We agree that feature-gated items which are "configured out" by should not generate warnings, whether they result from macro expansion or not. i.e., this PR seems good. :)
